algorithms/HPdclassifier/R/predict.hpdegbm.R

```diff
@@ -0,0 +1,218 @@
+#  GNU General Public License for more details.
+#
+#  A copy of the GNU General Public License is available at
+#  http://www.r-project.org/Licenses/
+####################################################################################
+
+
+####################################################################################
+###                     Data-distributed and model-centralized prediction function of distributed GBM
+####################################################################################
+predict.hpdegbm <- function(object, newdata, appType="binary-classification", type="link", trace = FALSE)
+{
+  # model: assembled GBM model
+  # best.iter: best iteration of sub-models
+  # newdata: testing data in dframe/darray
+  # appType: regression, binary-classification, multi-classification
+  # type="link" or "response"
+
+  ############### data-distributed and model-centralized  prediction #################
+  # suitable for models whose size can fit into the memory
+  # train data: npartition_train/nExecutor_train
+  # test data : npartition_test/nExecutor_test. maybe different from train data
+
+  # check function arguments
+  if(missing(object))
+     stop("'object' is a required argument")
+
+  if(missing(best.iter))
+     stop("'best.iter' is a required argument")
+
+  if(missing(newdata))
+     stop("'newdata' is a required argument")
+
+  if(!is.dframe(newdata) && !is.darray(newdata) && !is.data.frame(newdata) && !is.matrix(newdata))
+     stop("'newdata' must be a dframe or darray or data.frame or matrix")
+ 
+  # extract GBM models and corresponding best iterations (n.trees)
+  model <- object[[1]]
+  best.iter <- object[[2]]
+
+  # prediction for distributed big data 
+  if((is.dframe(newdata)) || (is.darray(newdata))) {
+     npartition_test <- npartitions(newdata)  
+     nExecutor_test  <- npartition_test
+
+     nTest <- nrow(newdata)
+     daPredict <- darray(dim=c(nTest,1), blocks=c(ceiling(nTest/npartition_test),1), sparse=FALSE)
+
+
+     # distributed prediction
+     foreach(i, 1:nExecutor_test, function(GBM_model=model, best.iter=best.iter, predi=splits(daPredict,i), data2=splits(newdata,i),appType=appType, type=type) { 
+        library(gbm)
+
+        # Gaussian distribution for regression
+        if (appType == "regression") {
+           # Prediction by the first sub-model: AdaBoost, Bernoulli distribution for binary classification
+           GBM_modeli <- GBM_model[[1]]
+           best.iteri <- best.iter[1,1]
+           predi <- predict(GBM_modeli, newdata, best.iteri, type="link")  
+
+           # fusion with other sub-models
+           npartition_train <- nrow(best.iter)
+           for (k in 2: npartition_train) {
+               GBM_modelk <- GBM_model[[k]]
+               best.iterk <- best.iter[k,1]
+               predi <- predi + predict(GBM_modelk, newdata, best.iterk, type="link")
+           } 
+           predi <- predi/npartition_train # average of regressions from sub-models
```

Comment:
why seperate predi for the k=1 case? initialize predi to be 0 and just run from k=1:npartition_train